### DISCUSSION TITLE
Accessing BoxPlayer object for non-rostered players in previous weeks

### DISCUSSION THREAD

**Discussion by johnkohler00** [cwendt94/espn-api#507] [ESPN API]
I want to get the actual and projected points for all players in every week, whether that player was rostered or not. I understand that the only place where the weekly projected points are accessible is the Box Player object. I thought that the `league.free_agents` call would work here but it only returns the _previous_ stats of _current_ free agents so players that have been picked up throughout the season do not show up here. Any ideas for how to access this data?

**Comment by cwendt94** [comment]
Right now there is not a direct call doing this but you can leverage the `league.player_map` which has all of players in the nfl mapping playerId to playerName. With this information you can go through all of the players and call `league.player_info(playerId=[1234, 456])` which will return a player class that has stats for each week with projected and actual. You will probably want to batch the player_info call and not add all of the playerIds as once.

**Reply by johnkohler00** [reply]
I think this strategy would work if I wanted season level stats (player.stats[0]['projected_points'] exists) but unfortunately I don't think the package returns *projected* stats for any particular week in the player objects (player.stats[1]['projected_points'] does not exist). I have only been able to find those dictionary entries in the stats objects attatched to 'BoxPlayer' objects. The Player class seems to only include actual stats rather than projected stats at the week-level.

**Reply by cwendt94** [reply]
It kind of depends where the data is being sourced. For the initial load of a team the player stats in the roster are just the season but when its being pulled directly from the `player_info` call it should break it down for each scoring period

**Reply by johnkohler00** [reply]
What am I missing here? I don't see the projected points. (This is for Christian McCaffery btw)
![Screenshot 2023-12-01 at 4 05 07 PM](https://github.com/cwendt94/espn-api/assets/44691145/6452c11e-4326-45cf-9034-676a6adf2f9d)

**Reply by cwendt94** [reply]
Yeah I think you are correct. I am looking at the API calls now and it seems like when you get just all the information for the player they don't have projected. Hmm Im not sure if there is a easy way to just get every week projected scores for a player.

**Reply by johnkohler00** [reply]
I see. So it is not possible to access the BoxPlayer class for a player in a given week if they were not rostered in that week in your league unless they are currently a free agent? It seems odd to me that the ESPN API gives us no way to pull the stats from previous weeks for some players (before they were picked up) that are currently rostered but they do give us a way to see those details as soon as someone drops that player...

**Reply by cwendt94** [reply]
Okay I played around with free agents logic a little and think I could adjust the call to get all players. This will only however get the projected stats of the week that is passed in the call `free_agents(self, week: int=None, size: int=50, position: str=None, position_id: int=None)`. So you would need to get week by week making the call. Im surprised they don't have a way with a single call to get all of these stats

**Reply by johnkohler00** [reply]
How were you able to play around with the free agents logic to get all players? I still can't get projected stats for certain players.

For example, in my league the Jaguars D/ST was a free agent until week 11 but I have no way of getting their box score objects for those earlier weeks because they are currently on a roster.
![IMG_7041](https://github.com/cwendt94/espn-api/assets/44691145/b7e3eab5-1bba-4d64-8cb1-466854b2b1b0)

**Reply by johnkohler00** [reply]
![Screenshot 2023-12-02 at 11 33 58 AM](https://github.com/cwendt94/espn-api/assets/44691145/bae9a3df-2cea-43b2-ba48-c4690cfb9fda)

**Reply by cwendt94** [reply]
In free agents I updated the call to remove `"filterStatus":{"value":["FREEAGENT","WAIVERS"]}` which then will return all players even when they are not on a team. However I feel like we should be able to leverage player_info call and get projected score possibly by passing a scoring period.

**Reply by johnkohler00** [reply]
Okay! Thank you so much for doing that. How can I access this new version of your code? I have been installing from pip, and at least as of my last comment I wasn't seeing that change reflected in the results of my call (per the screenshot).

**Reply by cwendt94** [reply]
This change has not been pushed yet. I have been just trying it out locally and figuring out the best way to get the full stats for a player.
